crates/ty_python_semantic/src/types/infer.rs

```diff
@@ -1641,6 +1637,30 @@ impl<'db> TypeInferenceBuilder<'db> {
                 return;
             }
 
+            let declared_ty = self.file_expression_type(returns);
+
+            let scope_id = self.index.node_scope(NodeWithScopeRef::Function(function));
+            if scope_id.is_generator_function(self.index) {
+                let inferred_return = if function.is_async {
+                    KnownClass::AsyncGeneratorType
+                } else {
+                    KnownClass::GeneratorType
```

Comment:
Should we have a TODO here for inferring these as generic, specialized with the actual returned type (in or after your generic protocols PR)?

I think that will require a bit of restructuring of this check, since it means we need to use `self.return_types_and_ranges` and actually construct an inferred return type for each return statement in the function, wrapped with the right kind of generator type.

Reply:
yeah, it's going to be icky. But I think it can be easily postponed until after the alpha -- my generic-protocol PR shouldn't break this, I don't think.